File: neet_code_150_cpp/math_geometry/detect_square.cpp
#include <algorithm>
#include <iostream>
#include <vector>
#include <unordered_map>

//#define DEBUG

// 模範解答
class DetectSquaresAns { 
    public:
        DetectSquaresAns() {}

        void add(std::vector<int> point) {
            points[point[0]][point[1]]++;
        }

        int count(std::vector<int> point) {
            int x1 = point[0];
            int y1 = point[1];

            int result = 0;

            for (auto x = points.begin(); x != points.end(); x++) {
                std::unordered_map<int, int> yPoints = x->second;
                for (auto y = yPoints.begin(); y != yPoints.end(); y++) {
                    int x3 = x->first;
                    int y3 = y->first;

                    // skip points on same x-axis or y-axis
                    if (abs(x3 - x1) == 0 || abs(x3 - x1) != abs(y3 - y1)) {
                        continue;
                    }

                    result += points[x3][y3] * points[x1][y3] * points[x3][y1];
                }
            }

            return result;
        }
    private:
        // {x -> {y -> count}}
        std::unordered_map<int, std::unordered_map<int, int>> points;
};

// Rustの模範解答より
// 答えが合わない
class DetectSquares {
    private:
        std::unordered_map<int, std::unordered_map<int, int>> counts;

    public:
        DetectSquares() {}

        void add(std::vector<int> point) {
            ++counts[point[0]][point[1]];
        }

        int count(std::vector<int> point) {
            int res = 0;
            int px = point[0], py = point[1];

            for(auto& [x, pair]: counts) {
                for(auto& [y, c]: pair) {
#ifdef DEBUG
                    std::cout << x << "," << y << std::endl;
#endif
                    if(std::abs(py - y) != std::abs(px - x) || x == px || y == py ) {
                        continue;
                    }

#ifdef DEBUG
                    std::cout << counts[x][py] << ", ";
                    std::cout << counts[px][y] << std::endl;
#endif

                    res += counts[x][py] * counts[px][y]; 
                }

            }
            return res;
        }

#ifdef DEBUG
        void show(void) {

            for(const auto& [x, y_num]: counts) {
                for(const auto& [y, num]: y_num) {
                    std::cout << x << "," << y <<": " << num << std::endl;
                }
            }
        }
#endif
};

int main(void) {
    DetectSquares p_1;
    p_1.add(std::vector{3, 10});
    p_1.add(std::vector{11, 2});
    p_1.add(std::vector{3, 2});
#ifdef DEBUG
    p_1.show();
#endif
    std::cout << p_1.count(std::vector{11, 10}) << std::endl; 
    std::cout << p_1.count(std::vector{14, 8}) << std::endl; 
    p_1.add(std::vector{11, 2});
    std::cout << p_1.count(std::vector{11, 10}) << std::endl; 

    DetectSquaresAns p_2;
    p_2.add(std::vector{3, 10});
    p_2.add(std::vector{11, 2});
    p_2.add(std::vector{3, 2});
    std::cout << p_2.count(std::vector{11, 10}) << std::endl; 
    std::cout << p_2.count(std::vector{14, 8}) << std::endl; 
    p_2.add(std::vector{11, 2});
    std::cout << p_2.count(std::vector{11, 10}) << std::endl; 
}

File: neet_code_150/src/math_and_geometry/detect_square.rs
use std::collections::HashMap;
// 解けなかった
#[derive(Debug)]
struct DetectSquares {
    point: Vec<(i32, i32)>,
}

impl DetectSquares {
    fn new() -> Self {
        DetectSquares { point: vec![] }
    }

    fn add(&mut self, point: Vec<i32>) {
        self.point.push((point[0], point[1]));
    }

    fn count(&self, point: Vec<i32>) -> i32 {
        let p_x = point[0];
        let p_y = point[1];
        let mut number = 0;
        for p in &self.point {
            let mut len_sq = 0;
            if p.1 == p_y {
                len_sq = p_y - p.1;
            }
        }

        number
    }
}

// 模範解答
struct DetectSquaresAns {
    point: Vec<(i32, i32)>,
    counts: HashMap<(i32, i32), i32>,
}
impl DetectSquaresAns {
    fn new() -> Self {
        Self {
            point: vec![],
            counts: HashMap::new(),
        }
    }

    fn add(&mut self, point: Vec<i32>) {
        let p = (point[0], point[1]);
        self.point.push(p);
        *self.counts.entry(p).or_default() += 1;
    }

    fn count(&self, point: Vec<i32>) -> i32 {
        let mut res = 0;
        let (px, py) = (point[0], point[1]);
        for (x, y) in self.point.iter() {
            if (py - y).abs() != (px - x).abs() || *x == px || *y == py {
                continue;
            }

            res +=
                self.counts.get(&(*x, py)).unwrap_or(&0) * self.counts.get(&(px, *y)).unwrap_or(&0);
        }

        res
    }
}

// pointはなしでcountsだけでもできるのでは?
// 今の所ACしない
struct DetectSquareB {
    counts: HashMap<(i32, i32), i32>,
}

impl DetectSquareB {
    fn new() -> Self {
        Self {
            counts: HashMap::new(),
        }
    }

    fn add(&mut self, point: Vec<i32>) {
        let p = (point[0], point[1]);
        *self.counts.entry(p).or_default() += 1;
    }

    fn count(&self, point: Vec<i32>) -> i32 {
        let mut res = 0;
        let (px, py) = (point[0], point[1]);
        for ((x, y), _c) in self.counts.iter() {
            if (py - y).abs() != (px - x).abs() || *x == px || *y == py {
                continue;
            }

            res +=
                self.counts.get(&(*x, py)).unwrap_or(&0) * self.counts.get(&(px, *y)).unwrap_or(&0);
        }

        res
    }
}

fn main() {
    let mut p_1 = DetectSquares::new();
    p_1.add(vec![3, 10]);
    p_1.add(vec![11, 2]);
    p_1.add(vec![3, 2]);

    p_1.count(vec![11, 10]);
    p_1.count(vec![14, 8]);

    p_1.add(vec![11, 2]);

    p_1.count(vec![11, 10]);

    let mut p_2 = DetectSquaresAns::new();
    p_2.add(vec![3, 10]);
    p_2.add(vec![11, 2]);
    p_2.add(vec![3, 2]);
    println!("{}", p_2.count(vec![11, 10]));
    println!("{}", p_2.count(vec![14, 8]));
    p_2.add(vec![11, 2]);
    println!("{}", p_2.count(vec![11, 10]));

    let mut p_3 = DetectSquareB::new();
    p_3.add(vec![3, 10]);
    p_3.add(vec![11, 2]);
    p_3.add(vec![3, 2]);
    println!("{}", p_3.count(vec![11, 10]));
    println!("{}", p_3.count(vec![14, 8]));
    p_3.add(vec![11, 2]);
    println!("{}", p_3.count(vec![11, 10]));
}
